### ISSUE TITLE
Severe glitches when hiding drag icon surface during dnd session

### ISSUE THREAD

**Issue by nickdiego** [swaywm/sway#5210]
I am these glitches while implementing Chromium's tab dragging, where I run a wayland dnd session the drag icon is used to emulate the "live window moving under the cursor". The issue happens when the window is snapped into a tab bar, which involves "hiding" the drag icon (to hide it I'm attaching a null buffer). Not sure if I'm doing something wrong or if it's just a bug in sway/wlroots, so appreciate your feedback/help on this.

**Sway version**: Tested with both 1.4 and latest master build
**Debug log**: [sway-dnd-icon-glitches.log](https://gist.github.com/nickdiego/84cb1d2be6d104d4af83ec13b75911cf)

*As a side note, the same impl runs quite smoothly in both Weston and Mutter.*

[![Watch the video](https://i.imgur.com/EVWOaQx.png)](https://youtu.be/WcwA1UF8FqU)

**Comment by nickdiego** [comment]
Did a few more tests and noticed that the issue doesn't happen if the data source is destroyed (when the window is snapped), instead of only setting null buffer to the drag icon surface. Does it imply that something is not being properly cleaned up when the icon surface content is removed [1] while the dnd session is still alive?

[1] https://gitlab.freedesktop.org/wayland/wayland/-/blob/master/protocol/wayland.xml#L1436-1437

**Comment by ascent12** [comment]
Yeah, it'll be a sway issue of some sort, either with damage tracking or some other thing related to drag icons. What you're describing sounds like a completely legal thing for a client to do.
Related: https://github.com/swaywm/sway/issues/4078

**Comment by nickdiego** [comment]
@ascent12 thanks for confirming.

Could you point me out where should I start looking at so I may try to dig into it a bit in my spare time?

**Comment by kennylevinsen** [comment]
@nickdiego Do you have a reproduction available?

This looks very much like a damage issue, possibly that sway used the wrong geometry to damage the icon when removing it, or possibly forgot to damage it on removal altogether.

**Comment by nickdiego** [comment]
Hi @kennylevinsen 

> @nickdiego Do you have a reproduction available?

Sorry I don't have it at hand for now, as it's still under development, and I don't think you'll want to compile Chromium just to try to repro this.

At some point, I was going to try to implement a minimal sample which could reproduce this but I haven't found time to do so yet :( Anyways, I guess it's pretty easily reproducible:
1. Start a dnd session, with a drag icon surface set;
2. attach some buffer to the drag icon surface
3. attach null buffer to the drag icon surface once the pointer hovers some specific location of a toplevel surface (eg: the surface that originated the DND session).

Please let me know if you need any other input to work on this.
thanks
